photon-api/src/main/scala/com/linkedin/photon/ml/transformers/GameTransformer.scala

```diff
@@ -0,0 +1,308 @@
+      "blocks are evicted). Useful for very large scoring tasks.")
+
+  //
+  // Initialize object
+  //
+
+  setDefaultParams()
+
+  //
+  // Parameter setters
+  //
+
+  def setInputColumnNames(value: InputColumnsNames): this.type = set(inputColumnNames, value)
+
+  def setModel(value: GameModel): this.type = set(model, value)
+
+  def setValidationEvaluators(value: Seq[EvaluatorType]): this.type = set(validationEvaluators, value)
+
+  def setLogDataAndModelStats(value: Boolean): this.type = set(logDataAndModelStats, value)
+
+  def setSpillScoresToDisk(value: Boolean): this.type = set(spillScoresToDisk, value)
+
+  //
+  // Params trait extensions
+  //
+
+  override def copy(extra: ParamMap): GameTransformer = {
+
+    val copy = new GameTransformer(sc, logger)
+
+    extractParamMap(extra).toSeq.foreach { paramPair =>
+      copy.set(copy.getParam(paramPair.param.name), paramPair.value)
+    }
+
+    copy
+  }
+
+  /**
+   * Set the default parameters.
+   */
+  private def setDefaultParams(): Unit = {
+
+    setDefault(inputColumnNames, InputColumnsNames())
+    setDefault(logDataAndModelStats, false)
+    setDefault(spillScoresToDisk, false)
+  }
+
+  /**
+   * Verify that the interactions between individual parameters are valid.
+   *
+   * @note In Spark, interactions between parameters are checked by
+   *       [[org.apache.spark.ml.PipelineStage.transformSchema()]]. Since we do not use the Spark pipeline API in
+   *       Photon-ML, we need to have this function to check the interactions between parameters.
+   *
+   * @throws IllegalArgumentException if a required parameter is missing or a validation check fails
+   */
+  protected[transformers] def validateParams(): Unit = {
+
+    // Just need to check that these parameters are explicitly set
+    getRequiredParam(model)
+  }
+
+  /**
+   * Return the user-supplied value for a required parameter. Used for mandatory parameters without default values.
+   *
+   * @tparam T The type of the parameter
+   * @param param The parameter
+   * @return The value associated with the parameter
+   * @throws IllegalArgumentException if no value is associated with the given parameter
+   */
+  private def getRequiredParam[T](param: Param[T]): T =
+    get(param)
+      .getOrElse(throw new IllegalArgumentException(s"Missing required parameter ${param.name}"))
+
+  //
+  // GameTransformer functions
+  //
+
+  /**
+   * Transform a [[DataFrame]] of data samples to include scores computed using the sample feature vectors and the
+   * currently set model.
+   *
+   * @param data Input [[DataFrame]] of samples
+   * @return Scored data samples
+   */
+  def transform(data: DataFrame): ModelDataScores = {
+
+    validateParams()
+
+    val randomEffectTypes = getRequiredParam(model)
+      .toMap
+      .values
+      .flatMap {
+        case rem: RandomEffectModel => Seq(rem.randomEffectType)
+        case _ => Seq()
```

Comment:
I think using `Option` would be more natural here, since we have either one element or none:

```scala
  .flatMap {
    case rem: RandomEffectModel => Some(rem.randomEffectType)
    case _ => None
  }
```